rowexec: fix and update BenchmarkJoinReader

asubiotto opened the pull request:
BenchmarkJoinReader previously attempted to benchmark a simple case. However,
the benchmark was set up incorrectly and no rows were being returned from the
storage layer because it was creating a transaction before any inserts had
happened.

This commit fixes the benchmark (and adds assertions) and introduces a varying
number of match ratios (how many rows are returned per lookup row) as well as
a benchmark for the case in which the lookup columns form a key and the
lookups can therefore be parallelized.

Disk spilling is not explicitly benchmarked, but some cases spill to disk and
the benchmark logs this. I think it's enough.

Release note: None (benchmark change)

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/48206)
<!-- Reviewable:end -->

yuzefovich reviewed (commented):
> Reviewed 2 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48206)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto)

___
*[pkg/sql/execinfra/testutils.go, line 82 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48206#-M6AUks-A2ka8n_aU22R:-M6AUks-A2ka8n_aU22S:b-1jro15) ([raw file](https://github.com/cockroachdb/cockroach/blob/13607cacd91a3db4a58d5d6ab4217920bb54960c/pkg/sql/execinfra/testutils.go#L82)):*
> ```Go
> func (r *RepeatableRowSource) ConsumerClosed() {}
> 
> type RowGeneratingSource struct {
> ```

nit: I think this should live in `rowexec` package since it is used only in there.
___
*[pkg/sql/execinfra/testutils.go, line 198 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48206#-M6AUwg1AC1Ccmmsh6yZ:-M6AUwg1AC1Ccmmsh6y_:bll500m) ([raw file](https://github.com/cockroachdb/cockroach/blob/13607cacd91a3db4a58d5d6ab4217920bb54960c/pkg/sql/execinfra/testutils.go#L198)):*
> ```Go
> 
> // RowDisposer is a RowReceiver that discards any rows Push()ed.
> type RowDisposer struct {
> ```

nit: this could also be moved out of `execinfra` into `rowexec` now (I think I put it here because we had `joinReader` and `indexJoiner` live in `execinfra` for some time).
___
*[pkg/sql/rowexec/joinreader_test.go, line 780 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48206#-M6AVybK1zUomqx0swnV:-M6AVybK1zUomqx0swnW:bh3wktd) ([raw file](https://github.com/cockroachdb/cockroach/blob/13607cacd91a3db4a58d5d6ab4217920bb54960c/pkg/sql/rowexec/joinreader_test.go#L780)):*
> ```Go
> 
> 			curValue := -1
> 			// capture matchesPerLookupRow for use in the generating function later
> ```

nit: `s/capture/Capture/g`.
___
*[pkg/sql/rowexec/joinreader_test.go, line 786 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48206#-M6AW6NU3Xjwc0gcHxnT:-M6AW6NU3Xjwc0gcHxnU:b-3t844l) ([raw file](https://github.com/cockroachdb/cockroach/blob/13607cacd91a3db4a58d5d6ab4217920bb54960c/pkg/sql/rowexec/joinreader_test.go#L786)):*
> ```Go
> 				idx := row - 1
> 				if idx%matchesPerLookupRow == 0 {
> 					// Increment curValue every columnDef.n values. The first value will be
> ```

nit: `s/columnDef.n/columnDef.matchesPerLookupRow/g`.
___
*[pkg/sql/rowexec/joinreader_test.go, line 824 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48206#-M6AWutu3wv5Icl6pHcY:-M6AWutu3wv5Icl6pHcZ:b-5sjjbm) ([raw file](https://github.com/cockroachdb/cockroach/blob/13607cacd91a3db4a58d5d6ab4217920bb54960c/pkg/sql/rowexec/joinreader_test.go#L824)):*
> ```Go
> 				b.Run(benchmarkName, func(b *testing.B) {
> 					tableName := tableSizeToName(rightSz)
> 					if !tableCreated {
> ```

It seems suspicious to me that we have `tableCreated` defined outside of for loops, but update it inside of the for loops. I feel like we want to have `createdTable[rightSz]`.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48206)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/rowexec/joinreader_test.go, line 824 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48206#-M6AWutu3wv5Icl6pHcY:-M6EuBS72QG2YqWXPa0R:bqmhy1h) ([raw file](https://github.com/cockroachdb/cockroach/blob/13607cacd91a3db4a58d5d6ab4217920bb54960c/pkg/sql/rowexec/joinreader_test.go#L824)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

It seems suspicious to me that we have `tableCreated` defined outside of for loops, but update it inside of the for loops. I feel like we want to have `createdTable[rightSz]`.
</blockquote></details>

This was leftover code. Changed to create a table outside the loops.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
<img class="emoji" title=":LGTM:" alt=":LGTM:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 13 of 13 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48206#-:-M6Fya8b6hLac0FF-DxD:bmxilld)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto)

___
*[pkg/sql/rowexec/joinreader_test.go, line 714 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48206#-M6FyROdCRQfmOnGog9F:-M6FyROdCRQfmOnGog9G:bkzy5xi) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec82302ccd2c64d6412cbb9d2fcc507331128307/pkg/sql/rowexec/joinreader_test.go#L714)):*
> ```Go
> // spill to disk, in which case the benchmark logs that the join spilled.
> func BenchmarkJoinReader(b *testing.B) {
> 	if testing.Short() {
> ```

How long does it take to run this benchmark?


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
bors r=yuzefovich

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48206#-:-M6Tj8gg-II7piNhbqtm:b-ewwdc)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/rowexec/joinreader_test.go, line 714 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48206#-M6FyROdCRQfmOnGog9F:-M6TjIdZ85KJ4vTNX8Qy:b-ag2995) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec82302ccd2c64d6412cbb9d2fcc507331128307/pkg/sql/rowexec/joinreader_test.go#L714)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

How long does it take to run this benchmark?
</blockquote></details>

Around 1-2 minutes.


<!-- Sent from Reviewable.io -->

asubiotto commented:
CI failure is fixed by #48336, unrelated to this PR.

asubiotto commented:
bors r=yuzefovich

Commit:
rowexec: fix and update BenchmarkJoinReader

BenchmarkJoinReader previously attempted to benchmark a simple case. However,
the benchmark was set up incorrectly and no rows were being returned from the
storage layer because it was creating a transaction before any inserts had
happened.

This commit fixes the benchmark (and adds assertions) and introduces a varying
number of match ratios (how many rows are returned per lookup row) as well as
a benchmark for the case in which the lookup columns form a key and the
lookups can therefore be parallelized.

Disk spilling is not explicitly benchmarked, but some cases spill to disk and
the benchmark logs this. I think it's enough.

Release note: None (benchmark change)

asubiotto commented:
bors r=yuzefovich
